File: js/test/sandbox_logs_release.test.ts
import { expect, test } from "vitest";
import { createServer } from "nice-grpc";
import {
  ModalClientDefinition,
  type SandboxGetLogsRequest,
  TaskLogsBatch,
} from "../proto/modal_proto/api";
import { ModalClient } from "modal";

/**
 * A V1 Sandbox's output comes from the control plane, not from the worker, so
 * none of the Sandbox connection's own release machinery covers it: the stream
 * is multiplexed onto the long-lived control-plane connection, which nothing
 * closes.
 *
 * The control plane here is a real server on a real connection on purpose. A
 * forgotten stream costs no socket and no handle of its own, so what a leak
 * keeps alive is the call itself — which only a server that counts its open
 * calls can see. A mock standing in for the stream has no call to leak, and a
 * test written against one passes whether or not the bug is there.
 */

/** V1-shaped, so the Sandbox reads its output through the control plane. */
const SANDBOX_ID = "sb-nGEijt9WbBMlGrsPH9FOaC";
const IDLE_TIMEOUT_MS = 400;
const LINE = "hello from the sandbox\n";

test("a pending output read stays active beyond the idle timeout", async () => {
  let sendOutput!: () => void;
  const outputGate = new Promise<void>((resolve) => {
    sendOutput = resolve;
  });
  const cp = await startFakeControlPlane(false, undefined, outputGate);
  try {
    const sandbox = await cp.client.sandboxes.fromId(SANDBOX_ID);
    const reader = sandbox.stdout.getReader();
    const pending = reader.read();
    expect(await waitFor(() => cp.live() === 1, 5000)).toBe(true);
    await new Promise((resolve) =>
      globalThis.setTimeout(resolve, 2 * IDLE_TIMEOUT_MS),
    );
    expect(cp.live()).toBe(1);
    expect(cp.started()).toBe(1);
    sendOutput();
    expect((await pending).value).toBe(LINE);
    expect(await waitFor(() => cp.live() === 0, 6 * IDLE_TIMEOUT_MS)).toBe(
      true,
    );
    await reader.cancel();
  } finally {
    sendOutput();
    await cp.shutdown();
  }
});

function mockJwt(): string {
  const header = btoa(JSON.stringify({ alg: "HS256", typ: "JWT" }));
  const exp = Math.floor(Date.now() / 1000) + 3600;
  return `${header}.${btoa(JSON.stringify({ exp }))}.signature`;
}

/**
 * A control plane that serves one batch per call and then holds the call open.
 * With `silent`, it sends nothing at all, standing in for a Sandbox that has
 * not printed anything yet.
 */
async function startFakeControlPlane(
  silent = false,
  authGate?: Promise<void>,
  outputGate?: Promise<void>,
) {
  let open = 0;
  let started = 0;
  const requestedCursors: string[] = [];
  let authenticating = false;

  const unimplemented: Record<string, unknown> = {};
  for (const methodName of Object.keys(ModalClientDefinition.methods)) {
    unimplemented[methodName] = () => {
      throw new Error(`${methodName} is not implemented in this control plane`);
    };
  }

  const server = createServer();
  server.add(ModalClientDefinition, {
    ...unimplemented,
    async authTokenGet() {
      authenticating = true;
      await authGate;
      return { token: mockJwt() };
    },
    async *sandboxGetLogs(
      request: SandboxGetLogsRequest,
      context: { signal: AbortSignal },
    ) {
      started++;
      requestedCursors.push(request.lastEntryId);
      open++;
      try {
        await outputGate;
        if (!silent) {
          const firstBatch = request.lastEntryId === "0-0";
          if (!firstBatch && request.lastEntryId !== "1-0") {
            throw new Error(`unexpected log cursor: ${request.lastEntryId}`);
          }
          yield TaskLogsBatch.create({
            entryId: firstBatch ? "1-0" : "2-0",
            items: firstBatch
              ? [{ data: LINE }, { data: LINE }]
              : [{ data: "subsequent output\n" }],
            eof: false,
          });
        }
        // Held open, as the control plane does while it waits for a Sandbox to
        // print something more, until the caller goes away.
        await new Promise<void>((resolve) => {
          context.signal.addEventListener("abort", () => resolve());
        });
      } finally {
        open--;
      }
    },
  } as never);

  const port = await server.listen("127.0.0.1:0");
  process.env["MODAL_SERVER_URL"] = `http://127.0.0.1:${port}`;
  const client = new ModalClient({
    tokenId: "test-token-id",
    tokenSecret: "test-token-secret",
  });

  client.profile.sandboxChannelIdleTimeoutMs = IDLE_TIMEOUT_MS;

  return {
    client,
    live: () => open,
    started: () => started,
    requestedCursors,
    authenticating: () => authenticating,
    shutdown: async () => {
      client.close();
      server.forceShutdown();
    },
  };
}

async function waitFor(
  condition: () => boolean,
  timeoutMs: number,
): Promise<boolean> {
  const deadline = Date.now() + timeoutMs;
  while (Date.now() < deadline) {
    if (condition()) return true;
    await new Promise((resolve) => globalThis.setTimeout(resolve, 25));
  }
  return condition();
}

test("cancelling a read that is waiting for output returns instead of hanging", async () => {
  const cp = await startFakeControlPlane(true);
  try {
    const sandbox = await cp.client.sandboxes.fromId(SANDBOX_ID);
    const reader = sandbox.stdout.getReader();

    // Nothing has been printed, so this read is parked on the wire.
    const pending = reader.read().catch(() => undefined);
    expect(await waitFor(() => cp.live() === 1, 5_000)).toBe(true);

    // Giving up has to reach the call itself. Unwinding the iterator cannot do
    // it on its own here: that waits for the read to settle, and the read is
    // waiting for output that is not coming.
    const cancelled = reader.cancel().then(() => "returned");
    const outcome = await Promise.race([
      cancelled,
      new Promise((resolve) =>
        globalThis.setTimeout(() => resolve("still waiting"), 10_000),
      ),
    ]);
    expect(outcome).toBe("returned");
    await pending;
  } finally {
    await cp.shutdown();
  }
});

test.each([false, true])(
  "V1 output remains readable after detach (partly read: %s)",
  async (partlyRead) => {
    const cp = await startFakeControlPlane();
    try {
      cp.client.profile.sandboxChannelIdleTimeoutMs = 0;
      const sandbox = await cp.client.sandboxes.fromId(SANDBOX_ID);
      let reader;
      if (partlyRead) {
        reader = sandbox.stdout.getReader();
        expect((await reader.read()).value).toBe(LINE);
      }
      sandbox.detach();
      reader ??= sandbox.stdout.getReader();
      expect((await reader.read()).value).toBe(LINE);
      expect(cp.live()).toBe(1);
      expect(cp.started()).toBe(1);
      await reader.cancel();
      expect(await waitFor(() => cp.live() === 0, 5000)).toBe(true);
    } finally {
      await cp.shutdown();
    }
  },
);

test("explicit cancellation during authentication opens no log call", async () => {
  let finishAuth!: () => void;
  const gate = new Promise<void>((resolve) => {
    finishAuth = resolve;
  });
  const cp = await startFakeControlPlane(true, gate);
  try {
    const sandbox = await cp.client.sandboxes.fromId(SANDBOX_ID);
    const reader = sandbox.stdout.getReader();
    const pending = reader.read();
    expect(await waitFor(cp.authenticating, 5000)).toBe(true);
    const cancelled = reader.cancel();
    finishAuth();
    await cancelled;
    await pending;
    expect(cp.started()).toBe(0);
  } finally {
    finishAuth();
    await cp.shutdown();
  }
});

test("reading part of a Sandbox's output and forgetting it releases the stream", async () => {
  const cp = await startFakeControlPlane();
  try {
    const sandbox = await cp.client.sandboxes.fromId(SANDBOX_ID);
    const reader = sandbox.stdout.getReader();
    expect((await reader.read()).value).toBe(LINE);
    expect(cp.live()).toBe(1);

    // Walk away: stop reading, without cancel and without detach.
    expect(await waitFor(() => cp.live() === 0, 6 * IDLE_TIMEOUT_MS)).toBe(
      true,
    );
  } finally {
    await cp.shutdown();
  }
});

test("a released output stream reopens when it is read again", async () => {
  const cp = await startFakeControlPlane();
  try {
    const sandbox = await cp.client.sandboxes.fromId(SANDBOX_ID);
    const reader = sandbox.stdout.getReader();
    expect((await reader.read()).value).toBe(LINE);

    expect(await waitFor(() => cp.live() === 0, 6 * IDLE_TIMEOUT_MS)).toBe(
      true,
    );

    // Finish the buffered batch before reopening the released call.
    expect((await reader.read()).value).toBe(LINE);
    expect((await reader.read()).value).toBe("subsequent output\n");
    expect(cp.started()).toBe(2);
    expect(cp.requestedCursors).toEqual(["0-0", "1-0"]);
    await reader.cancel();
  } finally {
    await cp.shutdown();
  }
});

/** Puts several messages in one batch, which all resume from one entry. */
async function startMultiItemControlPlane() {
  const requestedCursors: string[] = [];

  const unimplemented: Record<string, unknown> = {};
  for (const methodName of Object.keys(ModalClientDefinition.methods)) {
    unimplemented[methodName] = () => {
      throw new Error(`${methodName} is not implemented in this control plane`);
    };
  }

  const server = createServer();
  server.add(ModalClientDefinition, {
    ...unimplemented,
    async authTokenGet() {
      return { token: mockJwt() };
    },
    async *sandboxGetLogs(
      request: SandboxGetLogsRequest,
      context: { signal: AbortSignal },
    ) {
      requestedCursors.push(request.lastEntryId);
      const firstBatch = request.lastEntryId === "0-0";
      if (!firstBatch && request.lastEntryId !== "1-0") {
        throw new Error(`unexpected log cursor: ${request.lastEntryId}`);
      }
      const lines = firstBatch ? ["a\n", "b\n", "c\n"] : ["d\n"];
      yield TaskLogsBatch.create({
        entryId: firstBatch ? "1-0" : "2-0",
        items: lines.map((data) => ({ data })),
        eof: false,
      });
      await new Promise<void>((resolve) => {
        context.signal.addEventListener("abort", () => resolve());
      });
    },
  } as never);

  const port = await server.listen("127.0.0.1:0");
  process.env["MODAL_SERVER_URL"] = `http://127.0.0.1:${port}`;
  const client = new ModalClient({
    tokenId: "test-token-id",
    tokenSecret: "test-token-secret",
  });
  client.profile.sandboxChannelIdleTimeoutMs = IDLE_TIMEOUT_MS;

  return {
    client,
    requestedCursors,
    shutdown: async () => {
      client.close();
      server.forceShutdown();
    },
  };
}

test("a release part way through a batch loses nothing", async () => {
  const cp = await startMultiItemControlPlane();
  try {
    const sandbox = await cp.client.sandboxes.fromId(SANDBOX_ID);
    const reader = sandbox.stdout.getReader();
    expect((await reader.read()).value).toBe("a\n");

    // Pause past the timeout, so the stream goes back mid-batch. Every message
    // in it resumes from the same entry, so reopening would step over the rest.
    await new Promise((resolve) =>
      globalThis.setTimeout(resolve, IDLE_TIMEOUT_MS * 3),
    );

    expect((await reader.read()).value).toBe("b\n");
    expect((await reader.read()).value).toBe("c\n");
    // Only then does it carry on from after that batch.
    expect((await reader.read()).value).toBe("d\n");
    expect(cp.requestedCursors).toEqual(["0-0", "1-0"]);
    await reader.cancel();
  } finally {
    await cp.shutdown();
  }
});

File: go/test/sandbox_logs_release_test.go
package test

import (
	"context"
	"net"
	"sync/atomic"
	"testing"
	"time"

	modal "github.com/modal-labs/modal-client/go"
	pb "github.com/modal-labs/modal-client/go/proto/modal_proto"
	"github.com/onsi/gomega"
	"go.uber.org/goleak"
	"google.golang.org/grpc"
	"google.golang.org/grpc/codes"
	"google.golang.org/grpc/credentials/insecure"
	"google.golang.org/grpc/status"
	"google.golang.org/grpc/test/bufconn"
)

// A V1 Sandbox's output comes from the control plane, not from the worker, so
// none of the Sandbox connection's own release machinery covers it: the stream
// sits on the long-lived control-plane connection, which nothing closes.
//
// The control plane here is a real server on a real connection on purpose.
// grpc-go runs a goroutine per client stream, and that goroutine is what a
// forgotten stream leaves behind; a hand-written stub standing in for the
// stream has no such goroutine, and a test written against one passes whether
// or not the leak is there.

// V1-shaped, so the Sandbox reads its output through the control plane.
const logsSandboxID = "sb-nGEijt9WbBMlGrsPH9FOaC"

type fakeLogsControlPlane struct {
	pb.UnimplementedModalClientServer
	line        string
	nextLine    string
	lastEntryID atomic.Value
	// When set, every call is failed with CANCELED instead of served, standing
	// in for a server that keeps cancelling.
	cancelEveryCall bool
	calls           atomic.Int64
	// Calls being served right now, which is what a leaked stream keeps alive.
	live atomic.Int64
}

// AppGetOrCreate gives the test a cheap call to open the connection with,
// before it starts counting goroutines.
func (s *fakeLogsControlPlane) AppGetOrCreate(
	_ context.Context, req *pb.AppGetOrCreateRequest,
) (*pb.AppGetOrCreateResponse, error) {
	return pb.AppGetOrCreateResponse_builder{AppId: req.GetAppName()}.Build(), nil
}

// SandboxGetLogs sends one batch and then holds the stream open, as the real
// control plane does while it waits for a Sandbox to print something more.
func (s *fakeLogsControlPlane) SandboxGetLogs(
	req *pb.SandboxGetLogsRequest, stream grpc.ServerStreamingServer[pb.TaskLogsBatch],
) error {
	s.calls.Add(1)
	s.lastEntryID.Store(req.GetLastEntryId())
	if s.cancelEveryCall {
		return status.Error(codes.Canceled, "cancelled by the server")
	}
	s.live.Add(1)
	defer s.live.Add(-1)
	entryID, line := "1-0", s.line
	switch req.GetLastEntryId() {
	case "0-0":
	case "1-0":
		entryID, line = "2-0", s.nextLine
	default:
		return status.Error(codes.InvalidArgument, "unexpected log cursor")
	}
	batch := pb.TaskLogsBatch_builder{
		EntryId: entryID,
		Items:   []*pb.TaskLogs{pb.TaskLogs_builder{Data: line}.Build()},
	}.Build()
	if err := stream.Send(batch); err != nil {
		return err
	}
	<-stream.Context().Done()
	return stream.Context().Err()
}

func startFakeLogsControlPlane(t *testing.T, line string) *modal.Client {
	t.Helper()
	client, _ := startFakeLogsControlPlaneWith(t, &fakeLogsControlPlane{line: line})
	return client
}

func startFakeLogsControlPlaneWith(t *testing.T, plane *fakeLogsControlPlane) (*modal.Client, *fakeLogsControlPlane) {
	t.Helper()
	g := gomega.NewWithT(t)

	listener := bufconn.Listen(1024 * 1024)
	server := grpc.NewServer()
	pb.RegisterModalClientServer(server, plane)
	go func() {
		if err := server.Serve(listener); err != nil {
			t.Logf("fake control plane stopped: %v", err)
		}
	}()

	conn, err := grpc.NewClient("passthrough:///bufnet",
		grpc.WithContextDialer(func(context.Context, string) (net.Conn, error) { return listener.Dial() }),
		grpc.WithTransportCredentials(insecure.NewCredentials()),
	)
	g.Expect(err).ToNot(gomega.HaveOccurred())

	client, err := modal.NewClientWithOptions(&modal.ClientParams{
		TokenID:            "test-token-id",
		TokenSecret:        "test-token-secret",
		Environment:        "test",
		ControlPlaneClient: pb.NewModalClientClient(conn),
		ControlPlaneConn:   conn,
	})
	g.Expect(err).ToNot(gomega.HaveOccurred())

	t.Cleanup(func() {
		client.Close()
		_ = conn.Close()
		server.Stop()
		_ = listener.Close()
	})
	return client, plane
}

// Closing a partly read V1 output stream must release its gRPC goroutine.
func TestSandboxLogsCloseLeavesNoGoroutinesBehind(t *testing.T) {
	g := gomega.NewWithT(t)

	const line = "hello from the sandbox\n"
	client := startFakeLogsControlPlane(t, line)

	// Open the connection first: its transport goroutines belong to the client,
	// not to the stream this test is watching.
	_, err := client.Apps.FromName(t.Context(), "warmup", &modal.AppFromNameParams{CreateIfMissing: true})
	g.Expect(err).ToNot(gomega.HaveOccurred())

	existing := goleak.IgnoreCurrent()

	sandbox, err := client.Sandboxes.FromID(t.Context(), logsSandboxID, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())

	buf := make([]byte, len(line))
	n, err := sandbox.Stdout.Read(buf)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(string(buf[:n])).To(gomega.Equal(line))

	g.Expect(sandbox.Stdout.Close()).To(gomega.Succeed())
	g.Eventually(func() error { return goleak.Find(existing) }, 8*time.Second, 50*time.Millisecond).
		Should(gomega.Succeed(), "closing a Sandbox output stream should release its goroutines")
}

// A server that keeps cancelling must exhaust the read's retry budget.
func TestServerCancellationDoesNotReopenForever(t *testing.T) {
	g := gomega.NewWithT(t)

	client, plane := startFakeLogsControlPlaneWith(t, &fakeLogsControlPlane{cancelEveryCall: true})
	sandbox, err := client.Sandboxes.FromID(t.Context(), logsSandboxID, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())

	read := make(chan error, 1)
	go func() {
		_, readErr := sandbox.Stdout.Read(make([]byte, 64))
		read <- readErr
	}()

	g.Eventually(read, 30*time.Second).Should(gomega.Receive(gomega.HaveOccurred()),
		"a read against a server that keeps cancelling should end in an error")
	// The retry budget is ten, so the call count must remain bounded.
	g.Expect(plane.calls.Load()).To(gomega.BeNumerically("<=", int64(20)))
}

// V1 output uses the shared control-plane client and remains usable after detach.
func TestDetachLeavesV1OutputReadable(t *testing.T) {
	t.Parallel()
	g := gomega.NewWithT(t)
	client, plane := startFakeLogsControlPlaneWith(t, &fakeLogsControlPlane{line: "hello\n"})
	sandbox, err := client.Sandboxes.FromID(t.Context(), logsSandboxID, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	t.Cleanup(func() { _ = sandbox.Stdout.Close() })

	buf := make([]byte, 3)
	_, err = sandbox.Stdout.Read(buf)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(string(buf)).To(gomega.Equal("hel"))
	g.Expect(sandbox.Detach()).To(gomega.Succeed())
	_, err = sandbox.Stdout.Read(buf)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(string(buf)).To(gomega.Equal("lo\n"))
	g.Expect(plane.live.Load()).To(gomega.Equal(int64(1)))
	g.Expect(sandbox.Stdout.Close()).To(gomega.Succeed())
	g.Eventually(plane.live.Load).Should(gomega.BeZero())

	unread, err := client.Sandboxes.FromID(t.Context(), logsSandboxID, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	t.Cleanup(func() { _ = unread.Stdout.Close() })
	g.Expect(unread.Detach()).To(gomega.Succeed())
	g.Expect(plane.calls.Load()).To(gomega.Equal(int64(1)))
	_, err = unread.Stdout.Read(buf)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(string(buf)).To(gomega.Equal("hel"))
	g.Expect(plane.calls.Load()).To(gomega.Equal(int64(2)))
}

// Reading part of a V1 Sandbox's output and then forgetting it must give the
// stream back, without a Close and without a Detach.
func TestSandboxLogsPartialReadLeavesNoGoroutinesBehind(t *testing.T) {
	g := gomega.NewWithT(t)

	const line = "hello from the sandbox\n"
	client := startFakeLogsControlPlane(t, line)

	// Open the connection first: its transport goroutines belong to the client,
	// not to the stream this test is watching.
	_, err := client.Apps.FromName(t.Context(), "warmup", &modal.AppFromNameParams{CreateIfMissing: true})
	g.Expect(err).ToNot(gomega.HaveOccurred())

	existing := goleak.IgnoreCurrent()

	sandbox, err := client.Sandboxes.FromID(t.Context(), logsSandboxID, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())

	buf := make([]byte, len(line))
	n, err := sandbox.Stdout.Read(buf)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(string(buf[:n])).To(gomega.Equal(line))

	// Walk away: no Close, no Detach. The stream is still open at this point,
	// and the idle release is what has to end it.
	g.Eventually(func() error { return goleak.Find(existing) }, 8*time.Second, 50*time.Millisecond).
		Should(gomega.Succeed(), "a Sandbox output stream nobody is reading should leave no goroutines behind")
}

// And the reader must still work afterwards: the next read reopens the stream
// and carries on from where the last one stopped.
func TestSandboxLogsReadAgainAfterIdleRelease(t *testing.T) {
	g := gomega.NewWithT(t)

	const line = "hello from the sandbox\n"
	const nextLine = "subsequent output\n"
	client, plane := startFakeLogsControlPlaneWith(t, &fakeLogsControlPlane{line: line, nextLine: nextLine})

	sandbox, err := client.Sandboxes.FromID(t.Context(), logsSandboxID, nil)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	t.Cleanup(func() { _ = sandbox.Stdout.Close() })

	buf := make([]byte, len(line))
	n, err := sandbox.Stdout.Read(buf[:5])
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(string(buf[:n])).To(gomega.Equal(line[:5]))
	g.Expect(plane.lastEntryID.Load()).To(gomega.Equal("0-0"))

	// Wait for idle cancellation before consuming the rest of the buffered batch.
	g.Eventually(plane.live.Load, 8*time.Second).Should(gomega.BeZero())
	n, err = sandbox.Stdout.Read(buf)
	g.Expect(err).ToNot(gomega.HaveOccurred())
	g.Expect(string(buf[:n])).To(gomega.Equal(line[5:]))
	g.Expect(plane.calls.Load()).To(gomega.Equal(int64(1)))

	// Read from a goroutine: a stream that was not given back leaves this read
	// waiting on a batch that never comes, and the test should say so rather
	// than hang until the package timeout.
	type readResult struct {
		text string
		err  error
	}
	done := make(chan readResult, 1)
	go func() {
		n, readErr := sandbox.Stdout.Read(buf)
		done <- readResult{text: string(buf[:n]), err: readErr}
	}()

	var got readResult
	g.Eventually(done, 10*time.Second).Should(gomega.Receive(&got),
		"a released stream should reopen on the next read")
	g.Expect(got.err).ToNot(gomega.HaveOccurred())
	g.Expect(got.text).To(gomega.Equal(nextLine))
	g.Expect(plane.lastEntryID.Load()).To(gomega.Equal("1-0"))
	g.Expect(plane.calls.Load()).To(gomega.Equal(int64(2)))

	g.Expect(sandbox.Stdout.Close()).To(gomega.Succeed())
}
